### DISCUSSION TITLE
API to generate runners token?

### DISCUSSION THREAD

**Discussion by colinwkirk** [community/community#26751] [Actions]
<p>It would be very useful to have an api endpoint that would allow me to generate setup tokens for self-hosted Actions Runners dynamically and allow me to have packerized AMIs that automagically pull the token and connect to github. The inability to do this currently drastically limits their usefulness for me.</p>

**Comment by Yanjingzhu** [comment]
<p>There is no Github API to generate tokens. When you want to setup self-hosted runner, you can go to repo Settings &gt; Actions &gt; Self-hosted runners section, clicking on “Add runner” button. Then there is a token for you to use in Configure steps.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/6693216b2b5321c337203f2fc72212ab36b5620f.png" data-download-href="/uploads/short-url/eDpVw8Zdl0Vzvmi3xOMiUklbl9l.png?dl=1" title="add runner.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181099744-c47196c5-8a72-481c-887b-da49651a48eb.png" alt="add runner.png" data-base62-sha1="eDpVw8Zdl0Vzvmi3xOMiUklbl9l" width="647" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/6/6693216b2b5321c337203f2fc72212ab36b5620f_2_647x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/6/6693216b2b5321c337203f2fc72212ab36b5620f_2_970x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/6/6693216b2b5321c337203f2fc72212ab36b5620f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/6693216b2b5321c337203f2fc72212ab36b5620f_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">add runner.png</span><span class="informations">1056×815 53.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>For more information of Adding self-hosted runners, please refer to <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fcn%2Factions%2Fautomating-your-workflow-with-github-actions%2Fadding-self-hosted-runners&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C30de9b82e63e43df2b8908d76f2c3374%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637100107939262593&amp;sdata=tEyjeiMMmH3j%2FxYqI%2B%2BVL0WFN5jWnhv8p2%2BdBjTnqGE%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/cn/actions/automating-your-workflow-with-github-actions/adding-self-hosted-runners</a></p>

**Comment by colinwkirk** [comment]
<p>I know there isn’t an api. That’s why I said it would be useful to have one.</p>

**Comment by Yanjingzhu** [comment]
<p>Sorry to tell you that as a security precaution, Github doesn’t provide an API to create personal access token. You need to follow the steps to create a PAT through web site.   <a href="https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line" rel="nofollow noopener">https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line</a></p>

**Comment by goughlee** [comment]
<p>So a normal PAT can be used for this purpose of registering a self hosted action runner? Does it need any specific permissions in order to do this?</p>

**Comment by colinwkirk** [comment]
<p>I don’t entirely understand how allowing api access to retrieve a token would be less secure than, say, using api access to interact with branches and repos and pull requests. There are already authenticated endpoints that plenty of automation users leverage.</p>

**Comment by goughlee** [comment]
<p>I’ve confirmed that PAT tokens cannot currently be used for this purpose, only the tokens obtained via Actions &gt; Add Runner work. Which is a bit of a problem for us as they appear to be short lived.</p>

**Comment by mscoutermarsh** [comment, marked as answer]
<p>An API for this is on the roadmap. I don’t have a timeline to share at the moment. But we’ll be posting to the Changelog when this is available.</p>
<p>And to clear up some of the confusion around PATs/runner tokens. The runner token provided via the UI is a temporary token that expires after 60 minutes. It only has the ability to register runners.</p>
<p>PATs are not able to register runners.</p>

**Comment by colinwkirk** [comment]
<p>Perfect, good to know. Thanks!</p>

**Comment by goughlee** [comment]
<p>Good to hear there will be an api to generate these tokens. We are trying to have self hosted action runners in aws fargate and so short lived tokens present a problem as the the runner tasks are transient. Also it would be benefitial to be able to scale them up and down as and when needed. Without the token api (or much longer lived tokens) this currently isn’t possible.</p>

**Comment by csumpter** [comment]
<p>Also looking forward to this feature. I was also looking to provision short lived runners, but cannot easily without the token generation API.</p>

**Comment by michaelgaida** [comment]
<p>Hey guys,</p>
<p>glad to here there is an API on the roadmap. Is there any ETA, so I know when about I can make the provisioning fully automated?</p>
<p>Thanks guys,</p>
<p>Michael</p>

**Comment by mkrakowitzer** [comment]
<p>I asked GitHub support for an ETA, should be sometime next week. \o/</p>

**Comment by chrispat** [comment]
<p>The GitHub Actions API has shipped <a href="https://developer.github.com/v3/actions/self_hosted_runners/" rel="nofollow noopener">https://developer.github.com/v3/actions/self_hosted_runners/</a></p>

**Comment by agunescu** [comment]
<p>The generated self-hosted using API, is 130 characters long, way longer then the one generated on</p>
<p>"Settings -&gt; Action -&gt; Add runner ", doesn’t seems to work either, any ideas why?</p>

**Comment by gajus** [comment]
<p>Is there a GraphQL API endpoint?</p>

**Comment by brianjmurrell** [comment]
Anyone successfully using the API with a token?  I'm getting a strange response:

```
$ curl -L   -H "Accept: application/vnd.github+json"   -H "Authorization: Bearer [redacted]"  -H "X-GitHub-Api-Version: 2022-11-28"   https://api.github.com/orgs/<my_org>/actions/runners
{
  "message": "Must have admin rights to Repository.",
  "documentation_url": "https://docs.github.com/rest/reference/actions#list-self-hosted-runners-for-an-organization"
}
```

But the operation I am trying to do, listing organization runners is not related to a repository but rather an organization's runners.  A repository is not even mentioned in the API request.

I've even gone overboard and allowed every scope possible:

![image](https://user-images.githubusercontent.com/559317/226422918-e4765429-d341-4cd6-861a-e965d1f53d24.png)

 in an effort to debug but still get the same error above.
